cmd/nitro/init.go

```diff
@@ -173,7 +186,221 @@ func validateBlockChain(blockChain *core.BlockChain, expectedChainId *big.Int) e
+	}
+	arbDb, err := stack.OpenDatabase("arbitrumdata", 0, 0, "", true)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		err := arbDb.Close()
+		if err != nil {
+			log.Warn("failed to close arbitrum database after finding pruning targets", "err", err)
+		}
+	}()
+	roots := importantRoots{
+		chainDb: chainDb,
+	}
+	genesisNum := chainConfig.ArbitrumChainParams.GenesisBlockNum
+	genesisHash := rawdb.ReadCanonicalHash(chainDb, genesisNum)
+	genesisHeader := rawdb.ReadHeader(chainDb, genesisHash, genesisNum)
+	if genesisHeader == nil {
+		return nil, errors.New("missing L2 genesis block header")
+	}
+	err = roots.addHeader(genesisHeader, false)
+	if err != nil {
+		return nil, err
+	}
+	if initConfig.Prune == "validator" {
+		if l1Client == nil {
+			return nil, errors.New("an L1 connection is required for validator pruning")
+		}
+		callOpts := bind.CallOpts{
+			Context:     ctx,
+			BlockNumber: big.NewInt(int64(rpc.FinalizedBlockNumber)),
+		}
+		rollup, err := staker.NewRollupWatcher(rollupAddrs.Rollup, l1Client, callOpts)
+		if err != nil {
+			return nil, err
+		}
+		latestConfirmedNum, err := rollup.LatestConfirmed(&callOpts)
+		if err != nil {
+			return nil, err
+		}
+		latestConfirmedNode, err := rollup.LookupNode(ctx, latestConfirmedNum)
+		if err != nil {
+			return nil, err
+		}
+		confirmedHash := latestConfirmedNode.Assertion.AfterState.GlobalState.BlockHash
+		confirmedNumber := rawdb.ReadHeaderNumber(chainDb, confirmedHash)
+		var confirmedHeader *types.Header
+		if confirmedNumber != nil {
+			confirmedHeader = rawdb.ReadHeader(chainDb, confirmedHash, *confirmedNumber)
+		}
+		if confirmedHeader != nil {
+			err = roots.addHeader(confirmedHeader, false)
+			if err != nil {
+				return nil, err
+			}
+
+			validatorDb := rawdb.NewTable(arbDb, arbnode.BlockValidatorPrefix)
+			lastValidated, err := staker.ReadLastValidatedFromDb(validatorDb)
+			if err != nil {
+				return nil, err
+			}
+			if lastValidated != nil {
+				lastValidatedHeader := rawdb.ReadHeader(chainDb, lastValidated.BlockHash, lastValidated.BlockNumber)
+				if lastValidatedHeader != nil {
+					err = roots.addHeader(lastValidatedHeader, false)
+					if err != nil {
+						return nil, err
+					}
+				} else {
+					log.Warn("missing latest validated block", "number", lastValidated.BlockNumber, "hash", lastValidated.BlockHash)
+				}
+			}
+		} else {
+			log.Warn("missing latest confirmed block", "hash", confirmedHash)
+		}
+	} else if initConfig.Prune == "full" {
+		if nodeConfig.Node.Validator.Enable || nodeConfig.Node.BlockValidator.Enable {
+			return nil, errors.New("refusing to prune to full-node level when validator is enabled (you should prune in validator mode)")
+		}
+	} else if hashListRegex.MatchString(initConfig.Prune) {
+		parts := strings.Split(initConfig.Prune, ",")
+		var roots []common.Hash
+		for _, part := range parts {
+			roots = append(roots, common.HexToHash(part))
+		}
+		return roots, nil
```

Comment:
even when the user supplies a list, we need to add genesis (can be done on geth side as well)